### PR TITLE
feat: Minimal op registration in isolate

### PR DESCRIPTION
**Pull request by bartlomieju**
Towards #2987 

This PR introduces `OpRegistry` structure in `Isolate`.

`OpRegistry` collects `OpHandlers` which are functions returning `CoreOp`.

This can be safely introduced to work with `deno_cli` because if you set dispatch manually via `set_dispatch` then ops registered on isolate are ignored.

~~I still have to get op id forwarding to JS.~~

CC @ry @piscisaureus

### CODE REVIEW

File: core/isolate.rs [hunk 1]
```diff
@@ -257,6 +262,27 @@ impl Isolate {
...
+    self.op_registry.register_op(name, op)
+  }
+
+  pub fn call_op(
+    &self,
+    op_id: OpId,
+    control: &[u8],
+    zero_copy_buf: Option<PinnedBuf>,
+  ) -> CoreOp {
+    if op_id == 0 {
```
Comment (bartlomieju) [inline core/isolate.rs]: Temporary hack to easily fetch op map without adding binding in `libdeno`

**Comment by bartlomieju** [comment]
```
$ cargo run --release --example deno_core_http_bench

third_party/wrk/mac/wrk -d 10s --latency http://127.0.0.1:4544
Running 10s test @ http://127.0.0.1:4544
  2 threads and 10 connections
  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency   121.98us   41.94us   2.00ms   93.58%
    Req/Sec    39.43k     2.08k   41.72k    82.18%
  Latency Distribution
     50%  114.00us
     75%  119.00us
     90%  142.00us
     99%  292.00us
  792454 requests in 10.10s, 38.54MB read
Requests/sec:  78458.19
Transfer/sec:      3.82MB
```

```
$  cargo run --release --example deno_core_http_bench -- --multi-thread

third_party/wrk/mac/wrk -d 10s --latency http://127.0.0.1:4544
Running 10s test @ http://127.0.0.1:4544
  2 threads and 10 connections
  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency   185.81us  613.64us  18.64ms   98.18%
    Req/Sec    36.48k     6.32k   43.68k    87.13%
  Latency Distribution
     50%  106.00us
     75%  142.00us
     90%  228.00us
     99%    1.22ms
  733098 requests in 10.10s, 35.66MB read
Requests/sec:  72578.50
Transfer/sec:      3.53MB
```

File: core/examples/http_bench.js [hunk 2]
```diff
@@ -16,43 +15,6 @@ const responseBuf = new Uint8Array(
...
-    opRegistry.push(this);
-  }
-
-  setOpId(opId) {
-    this.opId = opId;
-  }
-
-  /** Returns i32 number */
-  sendSync(arg, zeroCopy = null) {
-    const buf = send(0, this.opId, arg, zeroCopy);
```
Comment (hayd) [inline core/examples/http_bench.js]: i think the idea behind this being a class method was the opId could be private (and perhaps not exposed)...

Comment (bartlomieju) [reply core/examples/http_bench.js]: Not really, not need to hide op id. Using class drastically lowers performance though.

File: core/ops.rs [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+use crate::CoreOpHandler;
+use crate::Op;
+use crate::OpId;
+use crate::PinnedBuf;
+use std::collections::HashMap;
+
+#[derive(Default)]
+pub struct OpRegistry {
+  pub ops: Vec<Box<CoreOpHandler>>,
+  pub phone_book: HashMap<String, OpId>,
```
Comment (kevinkassimo) [inline core/ops.rs]: 100% bikeshedding but maybe we should either leave comment on this or rename `phone_book` to something else 😅

File: core/shared_queue.js [hunk 4]
```diff
@@ -189,7 +195,9 @@ SharedQueue Binary Layout
       push,
       reset,
       shift
-    }
+    },
+    opsMap: {},
```
Comment (kevinkassimo) [inline core/shared_queue.js]: The exposure of `Deno.core.opsMap` to public revision might be a security issue? E.g. people could sabotage by swapping `mkdir` and `remove` op IDs (our current permission prompt only warns about `read/write/net` attempts but not identifiable op names)

Honestly I'm not sure if this is actually really related to PR though. I start to feel that exposure of the message passing channel (aka `Deno.core`) is a potentially dangerous operation, and access to it might require privileges. (hiding `Deno` and `core` should work under careful handling, as seen in `Worker` `noDenoNamespace` case (where `setTimeout` still privately uses `Deno.core` features))

**Comment by kevinkassimo** [comment]
> ```
> $ cargo run --release --example deno_core_http_bench
> 
> third_party/wrk/mac/wrk -d 10s --latency http://127.0.0.1:4544
> Running 10s test @ http://127.0.0.1:4544
>   2 threads and 10 connections
>   Thread Stats   Avg      Stdev     Max   +/- Stdev
>     Latency   121.98us   41.94us   2.00ms   93.58%
>     Req/Sec    39.43k     2.08k   41.72k    82.18%
>   Latency Distribution
>      50%  114.00us
>      75%  119.00us
>      90%  142.00us
>      99%  292.00us
>   792454 requests in 10.10s, 38.54MB read
> Requests/sec:  78458.19
> Transfer/sec:      3.82MB
> ```
> 
> ```
> $  cargo run --release --example deno_core_http_bench -- --multi-thread
> 
> third_party/wrk/mac/wrk -d 10s --latency http://127.0.0.1:4544
> Running 10s test @ http://127.0.0.1:4544
>   2 threads and 10 connections
>   Thread Stats   Avg      Stdev     Max   +/- Stdev
>     Latency   185.81us  613.64us  18.64ms   98.18%
>     Req/Sec    36.48k     6.32k   43.68k    87.13%
>   Latency Distribution
>      50%  106.00us
>      75%  142.00us
>      90%  228.00us
>      99%    1.22ms
>   733098 requests in 10.10s, 35.66MB read
> Requests/sec:  72578.50
> Transfer/sec:      3.53MB
> ```

Do you have the comparison with the master branch?

File: core/shared_queue.js (hunk 4 above)
Comment (bartlomieju) [reply core/shared_queue.js]: It should definitely be read-only for users, but we still need to ability to update this map (eg. for plugin registration).

**Comment by bartlomieju** [comment]
@kevinkassimo 

`master`
```
"req_per_sec": {
    "deno_tcp": 53518, 
    "deno_core_single": 54066, 
    "node_proxy_tcp": 20393, 
    "deno_proxy_tcp": 18601, 
    "deno_core_multi": 58185, 
    "deno_tcp_current_thread": 51059, 
    "hyper": 63312, 
    "node_tcp": 54550, 
    "deno_http": 18684, 
    "node_http": 19992, 
    "deno_proxy": 990, 
    "node_proxy": 2640
  }, 
  "max_latency": {
    "deno_tcp": 9.04, 
    "deno_core_single": 0.36, 
    "node_proxy_tcp": 1.05, 
    "deno_proxy_tcp": 5.36, 
    "deno_core_multi": 0.95, 
    "deno_tcp_current_thread": 9.46, 
    "hyper": 2.86, 
    "node_tcp": 0.597, 
    "deno_http": 2.62, 
    "node_http": 2.22, 
    "deno_proxy": 19.51, 
    "node_proxy": 9.4
  }, 
```

`this PR`
```
"req_per_sec": {
    "deno_tcp": 51090, 
    "deno_core_single": 58520, 
    "node_proxy_tcp": 19952, 
    "deno_proxy_tcp": 18476, 
    "deno_core_multi": 61798, 
    "deno_tcp_current_thread": 48471, 
    "hyper": 63395, 
    "node_tcp": 53003, 
    "deno_http": 18526, 
    "node_http": 19925, 
    "deno_proxy": 961, 
    "node_proxy": 2533
  }, 
  "max_latency": {
    "deno_tcp": 8.89, 
    "deno_core_single": 0.405, 
    "node_proxy_tcp": 1.06, 
    "deno_proxy_tcp": 5.62, 
    "deno_core_multi": 1.07, 
    "deno_tcp_current_thread": 9.13, 
    "hyper": 2.56, 
    "node_tcp": 0.588, 
    "deno_http": 2.72, 
    "node_http": 2.28, 
    "deno_proxy": 18.11, 
    "node_proxy": 10.16
  }, 
```

That's a nice speed up in req/secs, but latency is slightly worse.

File: core/ops.rs [hunk 5]
```diff
@@ -0,0 +1,47 @@
+// Copyright 2018-2019 the Deno authors. All rights reserved. MIT license.
+use crate::CoreOp;
+use crate::CoreOpHandler;
+use crate::Op;
+use crate::OpId;
```
Comment (ry) [inline core/ops.rs]: Maybe move the above declarations to this file

File: core/shared_queue.js [hunk 6]
```diff
@@ -178,6 +178,12 @@ SharedQueue Binary Layout
     return Deno.core.send(opId, control, zeroCopy);
   }
 
+  function refreshOpsMap() {
+    const opsMapBytes = dispatch(0, []);
+    const opsMapJson = String.fromCharCode.apply(null, opsMapBytes);
+    Deno.core.opsMap = JSON.parse(opsMapJson);
```
Comment (ry) [inline core/shared_queue.js]: Maybe just call this Deno.core.ops ?

File: core/isolate.rs (hunk 1 above)
Comment (ry) [reply core/isolate.rs]: I don’t understand your comment? Can you elaborate?

Comment (bartlomieju) [reply core/isolate.rs]: Ideally we want to register `get_ops_map` as an op using `register_op` method. 
```rust
...
let mut op_registry = OpRegistry::new();
op_registry.register_op("get_ops_map", |_control, _zero_copy| {
  // can't pass borrow checker
  let op_map = op_registry.get_op_map();
  let op_map_json = serde_json::to_string(&op_map).unwrap();
  let buf = op_map_json.as_bytes().to_owned().into_boxed_slice();
  Op::Sync(buf)
})
... 

Isolate {
...
op_registry,
}
```

This method would be placed in `op_registry` and reference that registry and I can't make types work

File: core/shared_queue.js (hunk 6 above)
Comment (bartlomieju) [reply core/shared_queue.js]: Removed completely. Please check out `Deno.core.initOps`

File: core/examples/http_bench.js [hunk 7]
```diff
@@ -1,11 +1,58 @@
 // This is not a real HTTP server. We read blindly one time into 'requestBuf',
 // then write this fixed 'responseBuf'. The point of this benchmark is to
 // exercise the event loop in a simple yet semi-realistic way.
-const OP_LISTEN = 1;
-const OP_ACCEPT = 2;
-const OP_READ = 3;
-const OP_WRITE = 4;
-const OP_CLOSE = 5;
+class Op {
```
Comment (bartlomieju) [inline core/examples/http_bench.js]: This class needs a bit more work. What I had in mind is that we create `JsonOp` which encapsulates json serialization/deserialization for ops. Then you create your own `JsonOp`s and during startup instead of calling `Deno.core.recv(...)` we'll call `Deno.core.initOps()` which will synchronize all ops between JS and Rust.

Comment (bartlomieju) [reply core/examples/http_bench.js]: Side note: Given that this is pretty elegant API IMHO that allows for "automagical" registration of ops, I'm still worried about performance of this solution because each op is effectively a closure. But let's see in benchmarks first.

File: core/isolate.rs (hunk 1 above)
Comment (bartlomieju) [reply core/isolate.rs]: For now I just used this hack and placed and registered a fake method in `OpRegistry` so ids are properly aligned

File: core/examples/http_bench.js (hunk 7 above)
Comment (bartlomieju) [reply core/examples/http_bench.js]: So I thought I nailed it, but this automagical registration might be harder to pull off than I expected. I got it working fine for `deno_core_http_bench`, but I can't get it working with CLI. The problem is snapshotting - ops are registered during snapshot build, but upon running new deno process shared queue is loaded during isolate startup and I guess the one from snapshot is ignored. 

For more detail see #3004

Comment (bartlomieju) [reply core/examples/http_bench.js]: I got another idea how to implement that, working on it

File: core/isolate.rs [hunk 8]
```diff
@@ -331,7 +336,11 @@ impl Isolate {
     let op = if let Some(ref f) = isolate.dispatch {
       f(op_id, control_buf.as_ref(), PinnedBuf::new(zero_copy_buf))
```
Comment (ry) [inline core/isolate.rs]: Since op_id == 0 has special meaning, we should assert in the case of the dispatch that we are not using it. Please add
```rust
assert!(op_id != 0);
```

File: core/isolate.rs (hunk 1 above)
Comment (ry) [reply core/isolate.rs]: Please add a comment here explaining that op_id 0 has special meaning to retrieve name => op_id map

Comment (ry) [reply core/isolate.rs]: Also add a comment in the triple slash rust doc of `Isolate::set_dispatch()` regarding the special meaning of op_id 0 and to `Isolate::register_op()`

File: core/isolate.rs [hunk 9]
```diff
@@ -257,6 +241,27 @@ impl Isolate {
     self.dispatch = Some(Arc::new(f));
   }
 
+  pub fn register_op(&mut self, name: &str, op: Box<CoreOpHandler>) -> OpId {
+    self.op_registry.register_op(name, op)
+  }
+
+  pub fn call_op(
```
Comment (ry) [inline core/isolate.rs]: I think this should be private

File: core/ops.rs [hunk 10]
```diff
@@ -0,0 +1,66 @@
...
+    self
+      .op_map
+      .entry(name.to_string())
+      .and_modify(|_| panic!("Op already registered {}", op_id))
+      .or_insert(op_id);
+
+    self.ops.push(serialized_op);
+    op_id
+  }
+}
```
Comment (ry) [inline core/ops.rs]: Add a unit test where you register an op, and assert that OpId is not zero.

Comment (ry) [reply core/ops.rs]: Also exercise `get_op_map()` in the unit tests here.

File: core/ops.rs [hunk 11]
```diff
@@ -0,0 +1,66 @@
...
+
+fn get_op_map(_control: &[u8], _zero_copy_buf: Option<PinnedBuf>) -> CoreOp {
+  Op::Sync(Box::new([]))
+}
+
+impl OpRegistry {
+  pub fn new() -> Self {
+    // TODO: this is make shift fix for get op map
+    let mut registry = Self::default();
+    registry.register_op("get_op_map", Box::new(get_op_map));
```
Comment (ry) [inline core/ops.rs]: assert return value is zero.

File: core/ops.rs [hunk 12]
```diff
@@ -0,0 +1,66 @@
...
+  pub op_map: HashMap<String, OpId>,
+}
+
+fn get_op_map(_control: &[u8], _zero_copy_buf: Option<PinnedBuf>) -> CoreOp {
+  Op::Sync(Box::new([]))
+}
+
+impl OpRegistry {
+  pub fn new() -> Self {
+    // TODO: this is make shift fix for get op map
```
Comment (ry) [inline core/ops.rs]: I don't understand the TODO. Can you rephrase it.

**Comment by ry** [comment]
@afinch7 your review would be appreciated if you have the time.

File: core/examples/http_bench.js [hunk 13]
```diff
@@ -1,11 +1,77 @@
...
+  static sendAsync(opId, arg, zeroCopy = null) {
+    const promiseId = nextPromiseId++;
+    const p = createResolvable();
+    promiseMap.set(promiseId, p);
+    send(promiseId, opId, arg, zeroCopy);
+    return p;
+  }
+}
+
+class HttpOp extends Op {
```
Comment (ry) [inline core/examples/http_bench.js]: What's the purpose of having these two classes in this small example? I suppose you're sketching out how it might work in the CLI, but I think this example should be as simple as possible. All ops are HttpOps. Please combine these two classes.

Comment (bartlomieju) [reply core/examples/http_bench.js]: That's right :+1: I'll merge them

File: core/examples/http_bench.js [hunk 14]
```diff
@@ -1,11 +1,77 @@
 // This is not a real HTTP server. We read blindly one time into 'requestBuf',
 // then write this fixed 'responseBuf'. The point of this benchmark is to
 // exercise the event loop in a simple yet semi-realistic way.
-const OP_LISTEN = 1;
-const OP_ACCEPT = 2;
-const OP_READ = 3;
-const OP_WRITE = 4;
-const OP_CLOSE = 5;
+const registry = {};
```
Comment (ry) [inline core/examples/http_bench.js]: I don't like that this registry map is a second source of truth for the op maps. (The real one lives in the "OpRegistry" in Rust.)
Can you remove this map entirely?

This is making me think maybe we need an op for adding ops.

File: core/examples/http_bench.js [hunk 15]
```diff
@@ -1,11 +1,77 @@
...
+  static handleAsyncMsgFromRust(opId, buf) {
+    const record = recordFromBuf(buf);
+    const { promiseId } = record;
+    const p = promiseMap.get(promiseId);
+    promiseMap.delete(promiseId);
+    p.resolve(record);
+  }
+
+  static sendSync(opId, arg, zeroCopy) {
+    const buf = send(0, opId, arg, zeroCopy);
```
Comment (ry) [inline core/examples/http_bench.js]: I don't like that op_id 0 has two different meanings...

File: core/examples/http_bench.js [hunk 16]
```diff
@@ -121,8 +163,8 @@ async function serve(rid) {
 }
 
 async function main() {
-  Deno.core.setAsyncHandler(handleAsyncMsgFromRust);
-
+  Deno.core.setAsyncHandler(Op.handleAsyncMsgFromRust);
+  initOps(Deno.core.getOps());
```
Comment (ry) [inline core/examples/http_bench.js]: I don't like all the indirection for simple logic like this. Remove the initOps function and put its body here.

File: core/examples/http_bench.js (hunk 14 above)
Comment (bartlomieju) [reply core/examples/http_bench.js]: It's not really a source of truth - Rust is source of truth in this case. This structure is used only to store all js ops in single place and set ops ids when they arrive from Rust on initialization. 

> This is making me think maybe we need an op for adding ops.

Can you elaborate?

File: core/examples/http_bench.js (hunk 15 above)
Comment (bartlomieju) [reply core/examples/http_bench.js]: How so? Only meaning is that op with id 0 is `get_op_map` op

File: core/examples/http_bench.js [hunk 17]
```diff
@@ -1,11 +1,77 @@
...
+    const res = await HttpOp.sendAsync(this.opId, arg, zeroCopy);
+    return res.result;
+  }
+}
+
+const OP_LISTEN = new HttpOp("listen");
+const OP_ACCEPT = new HttpOp("accept");
+const OP_READ = new HttpOp("read");
+const OP_WRITE = new HttpOp("write");
+const OP_CLOSE = new HttpOp("close");
```
Comment (ry) [inline core/examples/http_bench.js]: It's not clear to me if op ids have been assigned at this point? It seems not?

Comment (bartlomieju) [reply core/examples/http_bench.js]: No, they're not. Op ids are assigned after explicitly getting them from Rust with `Deno.core.getOps()`.

File: core/examples/http_bench.rs [hunk 18]
```diff
@@ -181,7 +156,11 @@ fn main() {
     });
 
     let mut isolate = deno::Isolate::new(startup_data, false);
-    isolate.set_dispatch(dispatch);
+    isolate.register_op("listen", serialize_http_op(op_listen));
+    isolate.register_op("accept", serialize_http_op(op_accept));
+    isolate.register_op("read", serialize_http_op(op_read));
+    isolate.register_op("write", serialize_http_op(op_write));
+    isolate.register_op("close", serialize_http_op(op_close));
```
Comment (ry) [inline core/examples/http_bench.rs]: I see, op_ids are assigned here... that answers [my previous question](https://github.com/denoland/deno/pull/3002/files#r328159510).

File: core/examples/http_bench.js (hunk 14 above)
Comment (ry) [reply core/examples/http_bench.js]: I think this registry map could be improved by using a Map type instead of an object and use js doc to label the type of key/values as @kitsonk has done [here](https://github.com/denoland/deno/blob/3d2d0ee771abe308aee9a0ab6c89fd09bac80330/deno_typescript/compiler_main.js#L112-L115).

>> This is making me think maybe we need an op for adding ops.

> Can you elaborate?

I'm not sure about it...  I will think about it more after we get this one landed.

File: core/examples/http_bench.js (hunk 15 above)
Comment (ry) [reply core/examples/http_bench.js]: Oh sorry, I thought the first argument of send() here was the op id. Ignore my comment.

File: core/isolate.rs (hunk 1 above)
Comment (bartlomieju) [reply core/isolate.rs]: I changed API a bit and moved all logic related to `ops` to `core/ops.rs`. So this `op_id === 0` situation is contained in that single file.

File: core/examples/http_bench.js [hunk 19]
```diff
@@ -121,7 +156,19 @@ async function serve(rid) {
...
-  Deno.core.setAsyncHandler(handleAsyncMsgFromRust);
+  Deno.core.setAsyncHandler(HttpOp.handleAsyncMsgFromRust);
+  // Initialize ops by getting their ids from Rust
+  // and assign id for each of our ops.
+  const opsMap = Deno.core.getOps();
+  for (const [name, opId] of Object.entries(opsMap)) {
+    const op = opRegistry.get(name);
+
+    if (!op) {
+      continue;
```
Comment (ry) [inline core/examples/http_bench.js]: Throw instead of continue

File: core/examples/http_bench.rs [hunk 20]
```diff
@@ -272,7 +267,13 @@ fn op_close(rid: i32) -> Box<HttpBenchOp> {
   }))
 }
 
-fn op_read(rid: i32, zero_copy_buf: Option<PinnedBuf>) -> Box<HttpBenchOp> {
```
Comment (ry) [inline core/examples/http_bench.rs]: Your patch includes some superfluous elements still, like changing the signature of this function.

Like another comment I made, I understand that you're sketching out how this would be in the more complex setting of `cli`, but I would nevertheless prefer that we keep this example as simple as possible. For example there shouldn't be a need for the `is_sync` parameter. All ops are know they are sync or async.

File: core/examples/http_bench.rs [hunk 21]
```diff
@@ -104,67 +98,48 @@ fn test_record_from() {
...
-        Ok(record_b)
-      })
-      .then(|result| -> Result<Buf, ()> {
-        let record = result.unwrap();
-        Ok(record.into())
-      }),
-  );
+) -> Box<HttpOp>;
+
+fn serialize_http_op(handler: HttpOpHandler) -> Box<CoreOpHandler> {
```
Comment (ry) [inline core/examples/http_bench.rs]: Just call this `http_op`

File: core/examples/http_bench.rs (hunk 20 above)
Comment (bartlomieju) [reply core/examples/http_bench.rs]: Done

File: core/isolate.rs [hunk 22]
```diff
@@ -244,19 +227,32 @@ impl Isolate {
...
   /// corresponds to the second argument of Deno.core.dispatch().
+  ///
+  /// If this method is used then ops registered using `op_register` function are
+  /// ignored and all dispatching must be handled manually in provided callback.
   pub fn set_dispatch<F>(&mut self, f: F)
   where
     F: Fn(OpId, &[u8], Option<PinnedBuf>) -> CoreOp + Send + Sync + 'static,
   {
     self.dispatch = Some(Arc::new(f));
   }
```
Comment (ry) [inline core/isolate.rs]: Is the plan to remove set_dispatch entirely? If so add a TODO.

File: core/isolate.rs [hunk 23]
```diff
@@ -244,19 +227,32 @@ impl Isolate {
...
 
+  /// New dispatch mechanism. Requires runtime to explicitly ask for op ids
+  /// before using any of the ops.
+  ///
+  /// Ops added using this method are only usable if `dispatch` is not set
+  /// (using `set_dispatch` method).
+  pub fn register_op(&mut self, name: &str, op: Box<CoreOpHandler>) -> OpId {
+    self.op_registry.register_op(name, op)
```
Comment (ry) [inline core/isolate.rs]: Add something like this:
```rust
assert!(self.dispatch.is_none(), "set_dispatch should not be used in conjunction with register_op");
```

File: core/ops.rs [hunk 24]
```diff
@@ -0,0 +1,115 @@
...
+  let buf = op_map_json.as_bytes().to_owned().into_boxed_slice();
+  Op::Sync(buf)
+}
+
+impl OpRegistry {
+  pub fn new() -> Self {
+    let mut registry = Self::default();
+    // Add single noop symbolizing "get_op_map" function. The actual
+    // handling is done in `call_op` method.
+    let op_id = registry.register_op("get_op_map", Box::new(op_noop));
```
Comment (ry) [inline core/ops.rs]: Why not `register_op("get_op_map", op_get_op_map)` ?

Comment (bartlomieju) [reply core/ops.rs]: See [this comment](https://github.com/denoland/deno/pull/3002#discussion_r326864872). Can't pass reference to `OpRegistry` and emplace this reference in the same registry and make borrow checker happy.

Comment (ry) [reply core/ops.rs]: Hmm... I can try to fix this in a bit. It seems like we should figure this out now. It should be possible.

Comment (hayd) [reply core/ops.rs]: could you move the Box::new into register_op?

File: core/ops.rs [hunk 25]
```diff
@@ -0,0 +1,115 @@
...
+    name: &str,
+    serialized_op: Box<CoreOpHandler>,
+  ) -> OpId {
+    let op_id = self.ops.len() as u32;
+
+    self
+      .op_map
+      .entry(name.to_string())
+      .and_modify(|_| panic!("Op already registered {}", op_id))
+      .or_insert(op_id);
```
Comment (ry) [inline core/ops.rs]: I think something like this would be more readable.
```rust
let existing = self.op_map.insert(name.to_string, op_id);
assert!(existing.is_none());
```

File: core/ops.rs (hunk 24 above)
Comment (bartlomieju) [reply core/ops.rs]: > could you move the Box::new into register_op?

I don't think so, compiler doesn't like passing unboxed functions

**Comment by ry** [comment]
Bartek - proxies shouldn’t be involved - they may kick us out of optimizations. I guess that Deno.ops[“foo”] should be an explicit function call instead. Maybe Deno.ops.get(“foo”)?

**Comment by bartlomieju** [comment]
> Bartek - proxies shouldn’t be involved - they may kick us out of optimizations. I guess that Deno.ops[“foo”] should be an explicit function call instead. Maybe Deno.ops.get(“foo”)?

Aaaah I was worried that's not the best solution. Yes, `Deno.ops.get("foo")` is the way to go, updating

File: core/isolate.rs [hunk 26]
```diff
@@ -329,9 +330,14 @@ impl Isolate {
     let isolate = unsafe { Isolate::from_raw_ptr(user_data) };
 
     let op = if let Some(ref f) = isolate.dispatch {
+      assert!(op_id != 0);
```
Comment (ry) [inline core/isolate.rs]:
```rust
assert!(op_id != 0, "op_id 0 is a special value that shouldn't be used with dispatch");
```

File: core/ops.rs [hunk 27]
```diff
@@ -0,0 +1,113 @@
...
+pub struct OpRegistry {
+  pub ops: Vec<Box<CoreOpHandler>>,
+  pub op_map: HashMap<String, OpId>,
+}
+
+fn op_noop(_control: &[u8], _zero_copy_buf: Option<PinnedBuf>) -> CoreOp {
+  Op::Sync(Box::new([]))
+}
+
+fn op_get_op_map(op_registry: &OpRegistry) -> CoreOp {
```
Comment (ry) [inline core/ops.rs]: This name strange. How about just `op_map()`?

File: core/shared_queue.js [hunk 28]
```diff
@@ -58,6 +59,24 @@ SharedQueue Binary Layout
     Deno.core.recv(handleAsyncMsgFromRust);
   }
 
+  function initOps() {
+    const opsMapBytes = Deno.core.send(0, new Uint8Array([]), null);
```
Comment (ry) [inline core/shared_queue.js]: Add comment about special meaning of op_id 0.

File: core/shared_queue.js [hunk 29]
```diff
@@ -189,6 +207,10 @@ SharedQueue Binary Layout
       push,
       reset,
       shift
+    },
+    ops: {
+      init: initOps,
+      get: getOpId
```
Comment (ry) [inline core/shared_queue.js]: `Deno.core.ops.get` is becoming distastefully long.

How about `Deno.core.ops()` returns the map. Remove the "init".
